### ISSUE TITLE
Performance degradation when using big tt-muncher `macro_rules`

### ISSUE THREAD

**Issue by Robbepop** [rust-lang/rust-analyzer#14101]
I recently noticed a big performance degradation regarding `rust-analyzer` in one of my projects: https://github.com/paritytech/wasmi

The problem is that `rust-analyzer` consumes much more memory than it has used to and also very often get stuck and shows incorrect warnings and won't finish analysis of the crate and therefore is pretty much unusable for the project. I noticed a very similar behavior of `rust-analyzer` in the https://github.com/bytecodealliance/wasm-tools repository.

After some backtracking I found that the performance regressions started after merging these PRs:
- https://github.com/paritytech/wasmi/pull/644
- https://github.com/paritytech/wasmi/pull/645

What do the PRs do?
They refactor use of `wasmparser`'s [`for_each_operator`](https://docs.rs/wasmparser/0.99.0/wasmparser/macro.for_each_operator.html) macro which is supposed to be used as tt-muncher macro in order to perform operations on the over 500 different Wasm operators.

Note that in particular in https://github.com/paritytech/wasmi/pull/645 we had to increase macro recursion limit in `lib.rs`:
```rust
#![recursion_limit = "750"]
```

**Comment by lnicola** [comment]
There are a couple of open issues about MBE performance. I suspect one of the problems is that we convert too often from token trees to a linear form.
